rfcs/text/0008-email.md

```diff
@@ -0,0 +1,144 @@
+# 0008: Email
+<!-- Leave this ID at 0000. The ECS team will assign a unique, contiguous RFC number upon merging the initial stage of this RFC. -->
+
+- Stage: **1** <!-- Update to reflect target stage. See https://elastic.github.io/ecs/stages.html -->
+- Date: **Oct 5th 2020** <!-- The ECS team sets this date at merge time. This is the date of the latest stage advancement. -->
+
+This RFC proposes a new top-level field to facilitate email use cases. 
+
+<!--
+As you work on your RFC, use the "Stage N" comments to guide you in what you should focus on, for the stage you're targeting.
+Feel free to remove these comments as you go along.
+-->
+
+<!--
+Stage 0: Provide a high level summary of the premise of these changes. Briefly describe the nature, purpose, and impact of the changes. ~2-5 sentences.
+-->
+
+## Fields
+
+<!--
+Stage 1: Describe at a high level how this change affects fields. Which fieldsets will be impacted? How many fields overall? Are we primarily adding fields, removing fields, or changing existing fields? The goal here is to understand the fundamental technical implications and likely extent of these changes. ~2-5 sentences.
+-->
+
+| field | type | description |
+| --- | --- | --- |
+| `email.action` | keyword | Action take by the source device, e.g. delivered, blocked, quarantined, deleted |
+| `email.bcc.address` | keyword | Addresses of Bcc's |
+| `email.bcc.domain` | keyword | Domains of the Bcc's |
+| `email.cc.address` | keyword | Addresses of Cc's |
+| `email.cc.domain` | keyword | Domains of Cc addresses |
+| `email.cipher` | keyword | Cipher used e.g. TLS |
+| `email.file.count` | value | Number of attachments included in the message |
+| `email.file.extension` | keyword | Extensions of attachment, e.g. .zip, .docx |
+| `email.file.hash` | keyword | Hash of attachments |
+| `email.file.name` | keyword | File name of attachements |
+| `email.file.size` | keyword | Total size of all attachements in bytes |
+| `email.direction` | keyword | Direction of the message based on the sending and receving domains |
+| `email.from.address` | keyword | Senders email address |
+| `email.from.domain` | keyword | Senders domain |
+| `email.latency` | keyword | The time, in milliseconds, the delivery attempt took |
+| `email.message_id` | keyword | Internet message ID of the message |
+| `email.process` | keyword | Name of the executable that carried out the transaction, e.g. outlook, sendmail |
+| `email.protocol` | keyword | The email protocol used, e.g. SMTP, IMAP |
+| `email.reply_to.address` | keyword | Reply-to address |
+| `object.return.address` | keyword | The return address for the message |
+| `email.size` | keyword | Total size of the message, in bytes, including attachments |
+| `email.subject` | keyword | Subject of the message |
+| `email.to` | keyword | Recipieint address |
```

Comment:
I think 'recipient' is a more common term when it comes to email than 'to'.
Also the schema should support multiple recipients.

Reply:
this really depends if you want to capture the semantics of the envelope/smtp and/or the email headers. This applies to other fields as well (to, cc, bcc, from)

Reply:
I agree there's nuances in recipients, and in how closely we want to represent each protocol.

After looking at a few raw emails, I'm starting to think we should keep the design of this field set pretty high level.

If capturing each protocol's nuances is desired or necessary, then I think we should consider working on a specific breakdown per protocol, as a separate step.

So I think the guiding principle we could use here is to capture the commonalities in `email.*`, even if it means merging some concepts, sometimes.

One point you're raising is pretty interesting, however. Should we capture each "type" of recipient in a different field, or in one parent field with an additional label to indicate which type of recipient this is?

The current proposal takes the former approach:

```
{ "email": {
  "to": [ {"address": "alice@example.com", ...} ],
  "cc": [ {"address": "bob@example.com", ...} ],
} }
```

The "recipient" suggestion would look like:

```
{ "email": {
  "recipients": [
    { "type": "to", "address": "alice@example.com", ...},
    { "type": "cc", "address": "bob@example.com", ...},
  ]
} }
```

🤔

Reply:
I really like this one:

`{ "email": {
  "to": [ {"address": "alice@example.com", ...} ],
  "cc": [ {"address": "bob@example.com", ...} ],
} }`

this avoid the need for nested objects. Also open to a new one in the "related" field:

`related.emails: ["alice@example.com", "bob@example.com"]`

Reply:
From an ingest perspective and a visualization perspective I think its better to keep the "to, cc, bcc" field structure over type.
The point from @vpiserchia is a good one, because once we start having list of objects we will have issues with the internal structure when it comes to visualizations, parsing and aggregations.